### PR TITLE
DOC: update Building from source docs for editable installs

### LINKED ISSUE #18900: BUG: Unable to use scipy development environment from IDE
### Describe your issue.

After following the steps to build `scipy` from source (https://scipy.github.io/devdocs/building/index.html#building-from-source):
```
mamba env create -f environment.yml
mamba activate scipy-dev
python dev.py build
```
I was unable to use the environment in the IDE (e.g. in the Jupyter Interactive Window in VSCode), as the `scipy` package cannot be imported:
```ModuleNotFoundError: No module named 'scipy'```

`scipy` is not present in the list of packages in the environment:
```conda list | grep scipy```

Presumably, this might be fixed once support for editable installs is added (https://github.com/scipy/scipy/issues/16293).

### Reproducing Code Example

```python
import scipy
```

### Error message

```shell
ModuleNotFoundError: No module named 'scipy'
```

### SciPy/NumPy/Python version and system information

```shell
1.12.0.dev0+1341.4172aee 1.25.1 sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /Users/artem/micromamba/envs/scipy-dev/include
    lib directory: /Users/artem/micromamba/envs/scipy-dev/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK=0 NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP=1 CORE2 MAX_THREADS=128
    pc file directory: /Users/artem/micromamba/envs/scipy-dev/lib/pkgconfig
    version: 0.3.23
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /Users/artem/micromamba/envs/scipy-dev/include
    lib directory: /Users/artem/micromamba/envs/scipy-dev/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK=0 NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP=1 CORE2 MAX_THREADS=128
    pc file directory: /Users/artem/micromamba/envs/scipy-dev/lib/pkgconfig
    version: 0.3.23
  pybind11:
    detection method: pkgconfig
    include directory: /Users/artem/micromamba/envs/scipy-dev/include
    name: pybind11
    version: 2.11.0
Compilers:
  c:
    args: -march=core2, -mtune=haswell, -mssse3, -ftree-vectorize, -fPIC, -fPIE, -fstack-protector-strong,
      -O2, -pipe, -isystem, /Users/artem/micromamba/envs/scipy-dev/include, -D_FORTIFY_SOURCE=2,
      -isystem, /Users/artem/micromamba/envs/scipy-dev/include
    commands: x86_64-apple-darwin13.4.0-clang
    linker: ld64
    linker args: -Wl,-pie, -Wl,-headerpad_max_install_names, -Wl,-dead_strip_dylibs,
      -Wl,-rpath,/Users/artem/micromamba/envs/scipy-dev/lib, -L/Users/artem/micromamba/envs/scipy-dev/lib,
      -march=core2, -mtune=haswell, -mssse3, -ftree-vectorize, -fPIC, -fPIE, -fstack-protector-strong,
      -O2, -pipe, -isystem, /Users/artem/micromamba/envs/scipy-dev/include, -D_FORTIFY_SOURCE=2,
      -isystem, /Users/artem/micromamba/envs/scipy-dev/include
    name: clang
    version: 15.0.7
  c++:
    args: -march=core2, -mtune=haswell, -mssse3, -ftree-vectorize, -fPIC, -fPIE, -fstack-protector-strong,
      -O2, -pipe, -stdlib=libc++, -fvisibility-inlines-hidden, -fmessage-length=0,
      -isystem, /Users/artem/micromamba/envs/scipy-dev/include, -D_FORTIFY_SOURCE=2,
      -isystem, /Users/artem/micromamba/envs/scipy-dev/include
    commands: x86_64-apple-darwin13.4.0-clang++
    linker: ld64
    linker args: -Wl,-pie, -Wl,-headerpad_max_install_names, -Wl,-dead_strip_dylibs,
      -Wl,-rpath,/Users/artem/micromamba/envs/scipy-dev/lib, -L/Users/artem/micromamba/envs/scipy-dev/lib,
      -march=core2, -mtune=haswell, -mssse3, -ftree-vectorize, -fPIC, -fPIE, -fstack-protector-strong,
      -O2, -pipe, -stdlib=libc++, -fvisibility-inlines-hidden, -fmessage-length=0,
      -isystem, /Users/artem/micromamba/envs/scipy-dev/include, -D_FORTIFY_SOURCE=2,
      -isystem, /Users/artem/micromamba/envs/scipy-dev/include
    name: clang
    version: 15.0.7
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.36
  fortran:
    args: -march=core2, -mtune=haswell, -ftree-vectorize, -fPIC, -fstack-protector,
      -O2, -pipe, -isystem, /Users/artem/micromamba/envs/scipy-dev/include
    commands: /Users/artem/micromamba/envs/scipy-dev/bin/x86_64-apple-darwin13.4.0-gfortran
    linker: ld64
    linker args: -Wl,-pie, -Wl,-headerpad_max_install_names, -Wl,-dead_strip_dylibs,
      -Wl,-rpath,/Users/artem/micromamba/envs/scipy-dev/lib, -L/Users/artem/micromamba/envs/scipy-dev/lib,
      -march=core2, -mtune=haswell, -ftree-vectorize, -fPIC, -fstack-protector, -O2,
      -pipe, -isystem, /Users/artem/micromamba/envs/scipy-dev/include
    name: gcc
    version: 12.2.0
  pythran:
    include directory: ../../../../micromamba/envs/scipy-dev/lib/python3.10/site-packages/pythran
    version: 0.13.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: darwin
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: darwin
Python Information:
  path: /Users/artem/micromamba/envs/scipy-dev/bin/python3.10
  version: '3.10'
```

### LINKED ISSUE #19565: DOC/DX: `meson-python` missing from 'required build dependencies'
### Issue with current documentation:

During investigation for gh-19563, I stumbled into this issue. In the 'Building from source to use SciPy' section of the [Building from source docs](http://scipy.github.io/devdocs/building/index.html#building-from-source), the 'required build dependencies' part is missing `meson-python`.

image

This results in an awfully confusing error message which refers to `mesonpy`, as @enn-nafnlaus encountered in https://github.com/scipy/scipy/issues/19153#issuecomment-1699525566.

<details>
<summary>Error Message</summary>

```
ERROR: Exception:
Traceback (most recent call last):
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/cli/base_command.py", line 180, in exc_logging_wrapper
    status = run_func(*args)
             ^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/cli/req_command.py", line 245, in wrapper
    return func(self, options, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/commands/install.py", line 377, in run
    requirement_set = resolver.resolve(
                      ^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 76, in resolve
    collected = self.factory.collect_root_requirements(root_reqs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 513, in collect_root_requirements
    reqs = list(
           ^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 474, in _make_requirements_from_install_req
    cand = self._make_candidate_from_link(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 211, in _make_candidate_from_link
    self._link_candidate_cache[link] = LinkCandidate(
                                       ^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 293, in __init__
    super().__init__(
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 156, in __init__
    self.dist = self._prepare()
                ^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 225, in _prepare
    dist = self._prepare_distribution()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 304, in _prepare_distribution
    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/operations/prepare.py", line 525, in prepare_linked_requirement
    return self._prepare_linked_requirement(req, parallel_builds)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/operations/prepare.py", line 640, in _prepare_linked_requirement
    dist = _get_prepared_distribution(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/operations/prepare.py", line 71, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/distributions/sdist.py", line 67, in prepare_distribution_metadata
    self.req.prepare_metadata()
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/req/req_install.py", line 577, in prepare_metadata
    self.metadata_directory = generate_metadata(
                              ^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/operations/build/metadata.py", line 35, in generate_metadata
    distinfo_dir = backend.prepare_metadata_for_build_wheel(metadata_dir)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_internal/utils/misc.py", line 772, in prepare_metadata_for_build_wheel
    return super().prepare_metadata_for_build_wheel(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_impl.py", line 186, in prepare_metadata_for_build_wheel
    return self._call_hook('prepare_metadata_for_build_wheel', {
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_impl.py", line 321, in _call_hook
    raise BackendUnavailable(data.get('traceback', ''))
pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 77, in _build_backend
    obj = import_module(mod_path)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lucascolley/mambaforge/envs/build-test/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1381, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1354, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1318, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'mesonpy'
```

</details>

The reasonable thing to try upon reading this error message is `mamba install mesonpy` / `pip install mesonpy`, which does not help at all.

### How to reproduce

```
mamba install python numpy cython pythran pybind11 compilers openblas pkg-config # as directed by the docs
git clone https://github.com/scipy/scipy.git
git submodule update --init
pip install . --no-build-isolation
```

### Idea or request for content:

Add `meson-python` into the quoted line in the docs.

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by rgommers**
This has been quite stable for some months, so it's time to add this as an option to the docs.

Also fixes a small issue with `meson-python` missing from listed dependencies.
    
Closes gh-18900
Closes gh-19565

### CODE REVIEW

File: doc/source/building/index.rst
Comment (DietBru) [inline doc/source/building/index.rst:393]: If the user has their pip-based virtual environment in a local ``venv`` or ``.venv`` directory, ``git clean -xdf`` will erase it, since these directories are listed in ``.gitignore``.  Also any local PyCharm customizations would be removed as well due to the ``.idea/`` entry in ``.gitignore``. Would it not  be sufficient to just erase the ``build`` and ``build-install`` directory?

File: doc/source/building/index.rst
Comment (mdhaber) [inline doc/source/building/index.rst:389]: Is there a possibility of updating `dev.py` to also work with development installs?

File: doc/source/building/index.rst
Comment (mdhaber) [inline doc/source/building/index.rst:387]:
```suggestion
    may use either an editable install or dev.py, but not both**. If you use
```

To make the rule easier to remember, maybe this can help the reader build a mental model of why.

The real limitation is that both can't be done within a single clone of the repository, right? It could be as simple as adding the qualification "in a given repository clone", but a separate sentence to explain would also work.

**Review by melissawm — APPROVED** [review]
This looks good to me, thanks @rgommers !

File: doc/source/building/index.rst
Comment (ev-br) [inline doc/source/building/index.rst:356]: Does it install `meson` itself through `meson-python` but not `ninja`?

Comment (rgommers) [reply doc/source/building/index.rst:356]: Yes indeed.

File: doc/source/building/index.rst
Comment (ev-br) [inline doc/source/building/index.rst:383]: Nit: strictly speaking, `pdb` breakpoints work just fine: drop a `breakpoint()` into the source, run `python dev.py test -t something/to/to/trigger/this/code/path` and you drop into pdb.
So it's some IDEs' problem?

File: doc/source/building/index.rst
Comment (ev-br) [inline doc/source/building/index.rst:388]: s/can/must ?

**Review by ev-br — APPROVED** [review]
LGTM modulo optional nits.

File: doc/source/building/index.rst
Comment (mdhaber) [inline doc/source/building/index.rst:382]: Without some trouble, it's also not easy to execute scripts using the "play" button of an IDE or use the IDE to run tests. (I forget whether I was ever successful.) I would suggest generalizing this statement of the limitation, maybe "because of the custom install location, SciPy installed using `dev.py` will only work via `dev.py`; for example, it will not be recognized within an IDE."

File: doc/source/building/index.rst
Comment (mdhaber) [inline doc/source/building/index.rst:402]:
```suggestion
    need to manually rebuild or reinstall. When you run ``git clean -xdf``, which removes
```
Since `pip install` looks like an install command.

File: doc/source/building/index.rst
Comment (mdhaber) [inline doc/source/building/index.rst:395]:
```suggestion
        # auto-rebuild each time SciPy is imported by the Python interpreter.
```

**Review by mdhaber — COMMENTED** [review]
Thanks for adding this!

File: doc/source/building/index.rst
Comment (rgommers) [reply doc/source/building/index.rst:383]: Yes, the problem is that in an IDE, you do something like clicking on the sidebar to set a breakpoint, and then the IDE knows how to tell its debugger that it's line X in file Y. Only works if you don't copy file Y ....

For CLI debuggers it's the same - works if you save the breakpoint in the file, not if you set it external to the file somehow.

File: doc/source/building/index.rst
Comment (rgommers) [reply doc/source/building/index.rst:389]: Yes, at some point. But I don't want to do that yet, because there are trade-offs. Things like multiple builds in parallel won't work, working with `__file__` may not work (not entirely sure), and error messages get confusing when you do `git clean -xdf` without `pip uninstal scipy`. So I'd say it's only for folks who really want the "breakpoint in IDE" support; for others `dev.py` as is is nicer.

File: doc/source/building/index.rst
Comment (rgommers) [reply doc/source/building/index.rst:387]: sure, I'll add that phrase. I don't really want to explain everything in detail, that will start duplication the meson-python docs.

**Review by mdhaber — APPROVED** [review]

**Comment by tylerjereddy** [comment]
I merged the doc improvement with 3 core dev approvals and passing CI, thanks.

**Pull request merged**
